File: Auth/Auth.Web/Configuration/OriginValidator.cs
using Microsoft.Extensions.Options;

namespace Auth.Web.Configuration
{
    public class OriginValidator(IOptions<OriginConfig> originConfig)
    {
        private readonly OriginConfig _originConfig = originConfig.Value;

        public bool IsAllowedOrigin(HttpRequest r)
        {
            var origin = GetOrigin(r);
            if(origin == null)
            {
                Error = "Empty origin received";
                return false;
            }

            var allowed = origin == _originConfig.Host;
            if(!allowed)
            {
                Error = $"Origin {origin} is not in the allowed list";
            }

            return allowed;
        }

        private string? GetOrigin(HttpRequest r)
        {
            var origin = r.Headers.Host.ToString()?.ToLower()?.Trim();
            if (string.IsNullOrEmpty(origin))
            {
                return null;
            }
            return origin;
        }

        public string Error { get; private set; } = string.Empty;
    }
}

File: NodeAuth/src/utils/originValidator.ts
import type { Request } from "express";

export type OriginValidationResult = {
  allowed: boolean;
  error: string;
  origin: string | null;
};

export const getRequestHost = (
  req: Pick<Request, "headers">,
): string | null => {
  const host = req.headers.host;
  if (!host) {
    return null;
  }
  if (Array.isArray(host)) {
    return host[0]?.toLowerCase().trim() || null;
  }
  return host.toLowerCase().trim();
};

export const validateOrigin = (
  req: Pick<Request, "headers">,
  allowedHost: string,
): OriginValidationResult => {
  const origin = getRequestHost(req);
  if (!origin) {
    return {
      allowed: false,
      error: "Empty origin received",
      origin: null,
    };
  }

  const allowed = origin === allowedHost.toLowerCase().trim();
  if (!allowed) {
    return {
      allowed: false,
      error: `Origin ${origin} is not in the allowed list`,
      origin,
    };
  }

  return {
    allowed: true,
    error: "",
    origin,
  };
};
